libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+	case !info.IsDir():
+		return fmt.Errorf("database directory path is not a directory: %s", dir)
+	}
+
+	return nil
+}
+
+// Scheme returns the scheme of the database.
+// This is only used in some API calls
+// and in StateDB to avoid iterating through deleted storage tries.
+// WARNING: If anything from go-ethereum uses this,
+// it must be overwritten to use something like:
+// `_, ok := db.(*Database); if !ok { return "" }`
+// to recognize the Firewood database.
+func (*Database) Scheme() string {
+	return rawdb.HashScheme
+}
+
+// Initialized checks whether a non-empty genesis block has been written.
+func (db *Database) Initialized(common.Hash) bool {
+	rootBytes, err := db.Firewood.Root()
+	if err != nil {
+		log.Error("firewood: error getting current root", "error", err)
+		return false
+	}
+	root := common.BytesToHash(rootBytes)
+	// If the current root isn't empty, then unless the database is empty, we have a genesis block recorded.
+	return root != types.EmptyRootHash
+}
+
+// Size returns the storage size of diff layer nodes above the persistent disk
+// layer and the dirty nodes buffered within the disk layer
+// Only used for metrics and Commit intervals in APIs.
+// This will be implemented in the firewood database eventually.
+// Currently, Firewood stores all revisions in disk and proposals in memory.
+func (*Database) Size() (common.StorageSize, common.StorageSize) {
+	return 0, 0
+}
+
+// This isn't called anywhere in coreth
+func (*Database) Reference(common.Hash, common.Hash) {}
+
+// Dereference drops a proposal from the database.
+// This function is no-op because unused proposals are dereferenced when no longer valid.
+// We cannot dereference at this call. Consider the following case:
+// Chain 1 has root A and root C
+// Chain 2 has root B and root C
+// We commit root A, and immediately dereference root B and its child.
+// Root C is Rejected, (which is intended to be 2C) but there's now only one record of root C in the proposal map.
+// Thus, we recognize the single root C as the only proposal, and dereference it.
+func (*Database) Dereference(common.Hash) {}
+
+// Firewood does not support this.
+func (*Database) Cap(common.StorageSize) error {
+	return nil
+}
+
+func (db *Database) Close() error {
+	db.proposalLock.Lock()
+	defer db.proposalLock.Unlock()
+
+	// All remaining proposals must be dereferenced.
+	db.possibleProposals = nil
+	db.proposalMap = nil
+	db.proposalTree = nil
+
+	// Close the database
+	// We must provide a context since it may hang while waiting for the finalizers to complete.
+	return db.Firewood.Close(context.Background())
+}
+
+func (db *Database) Update(root, parent common.Hash, block uint64, _ *trienode.MergedNodeSet, _ *triestate.Set, opts ...stateconf.TrieDBUpdateOption) error {
+	// We require block hashes to be provided for all blocks in production.
+	// However, many tests cannot reasonably provide a block hash for genesis, so we allow it to be omitted.
```

Comment:
Production code shouldn't be hobbled to allow for improper tests. Why can't a genesis block's hash be provided?